app/menus/TableOfContentsMenu.js

```diff
@@ -0,0 +1,60 @@
+// @flow
+
+import { observer } from "mobx-react";
+import { TableOfContentsIcon } from "outline-icons";
+import * as React from "react";
+import { useTranslation } from "react-i18next";
+import { MenuButton, useMenuState } from "reakit/Menu";
+import Button from "components/Button";
+import ContextMenu from "components/ContextMenu";
+import Template from "components/ContextMenu/Template";
+
+type Props = {|
+  headings: { title: string, level: number, id: string }[],
+|};
+
+function TableOfContentsMenu({ headings }: Props) {
+  const menu = useMenuState({
+    modal: true,
+    unstable_preventOverflow: true,
+    unstable_fixed: true,
+    unstable_flip: true,
+  });
+
+  const { t } = useTranslation();
+
+  const minHeading = headings.reduce(
+    (memo, heading) => (heading.level < memo ? heading.level : memo),
+    Infinity
+  );
+
+  return (
+    <>
+      <MenuButton {...menu}>
+        {(props) => (
+          <Button
+            {...props}
+            icon={<TableOfContentsIcon />}
+            iconColor="currentColor"
+            borderOnHover
+            neutral
+          />
+        )}
+      </MenuButton>
+      <ContextMenu {...menu} aria-label={t("Table of contents")}>
+        <Template
+          {...menu}
+          items={headings.map((heading) => {
```

Comment:
One last thing – can we add a "Table of contents" header to this menu, it's not immediately obvious what it represents especially if you're not familiar with the document content. There's a menu item type for this:

https://github.com/outline/outline/blob/f4959f98eb94713aa2adf190a0d56b57ee7a3dd3/app/components/ContextMenu/Template.js#L52-L54

Reply:
Yes Yes 😄

Reply:
It seems like this heading case is not handled in the code.

Reply:
This is how it looks now 

![Screen Shot 2021-07-09 at 13 49 22](https://user-images.githubusercontent.com/40148448/125047358-8385f580-e0bc-11eb-9012-565f3b7362ce.png)